DIPs/DIP1036.md

```diff
@@ -26,469 +28,303 @@ embedding arguments in the string itself.
 
-The spec can be used as follows.
+```D
+void foo(string s);
 
-`spec.toFormatString!(defaultSpec)` produces a compile-time format string with all the interpolated strings replaced as defined after the grammar below, but summarized as:
-1. If the prefix `{...}` is put between the `$` and the parameter, the token(s) between the `{` `}` is used.
-2. Otherwise, `defaultSpec` is used.
+foo(i"Hello, my name is ${name} and I am ${age} years old.");
 
-`spec.hasAllSpecs` is a compile-time boolean which indicates whether all interpolation parameters had specs defined with a `{}` prefix.
+string s = i"${var1} + ${var2} = ${var1 + var2}";
+```
 
-`spec` will automatically convert to a compile-time NUL-terminated C string if `spec.hasAllSpecs` is true. This allows it to be used for existing C functions such as `printf`.
+The above usage is translated directly to a compatible string and passed into
+the function, or assigned to the string variable. The library takes care of the
+translation, and provides a "best effort" attempt to translate any expression
+into character data.
 
-Example:
+The second case is for functions which can process the data without needing a string translation. For example:
 
 ```D
-void foo(T, Args...)(T spec, Args)
-if (isInterpolationSpec!T) // used for overloading with other functions
-{
-    import std.stdio;
-    string fmt = spec.toFormatString!"%s"; // also can be called at compile-time.
-    writeln(`D format spec: "`, fmt, `"`);
-    static if (spec.hasAllSpecs) // if all specs are specified
-    {
-        immutable char *zFormat = spec; // automatically converts to a C null-terminated string.
-        writeln(`C format spec: "`, zFormat[0 .. strlen(zFormat)], `"`);
-    }
-}
-
-void main()
-{
-    int apples = 5;
-    int bananas = 6;
-
-    foo(i"I ate $apples and ${%d}bananas totalling $(apples + bananas) fruit.");
-    // output:
-    // D format spec: "I ate %s and %d totalling %s fruit."
-
-    foo(i"I ate ${%d}apples and ${%d}bananas totalling ${%d}(apples + bananas) fruit.");
-    // output:
-    // D format spec: "I ate %d and %d totalling %d fruit."
-    // C format spec: "I ate %d and %d totalling %d fruit."
-}
+writeln(i"I ate ${apples} apples and ${bananas} bananas totalling ${apples + bananas} fruit.");
 ```
-Example of `printf` usage:
+
+is rewritten as:
 
 ```D
-printf(i"I ate ${%d}apples and ${%d}bananas totalling ${%d}(apples + bananas) fruit.\n");
+writeln(interp!"I ate "(), apples, interp!" apples and "(), bananas, interp!" bananas totalling "(),
+        apples + bananas, interp!" fruit."());
 ```
-This becomes (after converting the spec to a NUL-terinated C string):
+
+The `interp` template is defined by the D runtime library as follows:
+
 ```D
-printf("I ate %d and %d totalling %d fruit.", bananas, apples + bananas);
+struct interp(string s) // duplicated for wstring and dstring
+{
+    static string toString() { return s; }
```

Comment:
Since it's guaranteed to be a compile-time constant, would it make sense for this to be `enum string toString = s;` instead?

Reply:
As Adam noted, toString might always be called as a function, i.e. `obj.toString()`, and I want to support as many existing functions as possible.

This should work at compile-time without issue anyway.